### PR TITLE
Polar

### PR DESCRIPTION
**Pull request by daniel-k-taylor**
Still needs tests and verifying all the UI looks correct for all his random effects.

### CODE REVIEW

File: data/card_definitions.json [hunk 1]
```diff
@@ -15025,6 +15025,392 @@
...
+				"destination": "self",
+				"amount": 1,
+				"and": {
+					"effect_type": "choice",
+					"choice": [
+						{ "effect_type": "advance", "amount": 2 },
+						{ "effect_type": "advance", "amount": 3 }
+					]
+				}
+			}
```
Comment (Mach131) [inline data/card_definitions.json:15384]: Missing a "hit: push or pull 1"

File: globals/card_definitions.gd [hunk 2]
```diff
@@ -679,7 +704,10 @@ func get_effect_type_text(effect, card_name_source : String = "", char_effect_pa
 		"pull_to_buddy":
 			effect_str += "Pull %s to %s" % [str(effect['amount']), effect['buddy_name']]
 		"push":
-			effect_str += "Push " + str(effect['amount'])
+			var extra_info = ""
+			if 'save_buddy_spaces_entered_as_strike_x' in effect and effect['save_buddy_spaces_entered_as_strike_x']:
+				extra_info = "\nSet X to the number %s the opponent is pushed onto" % effect['buddy_name']
```
Comment (Mach131) [inline globals/card_definitions.gd:709]: number "of" %s ?

File: scenes/game/local_game.gd [hunk 3]
```diff
@@ -1396,6 +1400,69 @@ class Player:
...
+			if not opposing_player.is_in_location(location):
+				if opposing_player.is_in_location(location - 1) or opposing_player.is_in_location(location + 1):
+					matching_buddies.append(buddy_id_to_index.keys()[i])
+		return matching_buddies
+
+	func count_buddies_between_opponent():
+		var opposing_player = parent._get_player(parent.get_other_player(my_id))
+		var count = 0
+		var started = false
+		for location in range(MinArenaLocation, MaxArenaLocation + 1):
```
Comment (Mach131) [inline scenes/game/local_game.gd:1444]: I think I defined "is_left_of_location" and similar functions, if that makes this logic any easier

File: scenes/game/local_game.gd [hunk 4]
```diff
@@ -4281,8 +4384,9 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
 			events += [create_event(Enums.EventType.EventType_Strike_EffectChoice, performing_player.my_id, 0, "EffectOption")]
 		"move_to_space":
 			var space = effect['amount']
+			var remove_buddies_encountered = effect['remove_buddies_encountered']
```
Comment (Mach131) [inline scenes/game/local_game.gd:4387]: does this need a default value? or is it only reachable through move_to_any_space

File: scenes/game/local_game.gd [hunk 5]
```diff
@@ -4434,6 +4547,91 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
...
+				decision_info.source = effect['buddy_name']
+				var and_effect = null
+				var and_with_bonus = null
+				if 'and' in effect:
+					and_effect = effect['and']
+					and_with_bonus = and_effect
+				if 'if_removed_effect' in effect:
+					var additional_and_effect = and_effect
+					and_with_bonus = effect['if_removed_effect']
+					and_with_bonus['and'] = additional_and_effect
```
Comment (Mach131) [inline scenes/game/local_game.gd:4578]: would 'and' being set to null in an effect cause any issues? might be better to only set this if the and effect is non-null

File: scenes/game/local_game.gd [hunk 6]
```diff
@@ -4708,6 +4906,163 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
...
+						for i in range(valid_new_positions.size() - 1, -1, -1):
+							var check_position = valid_new_positions[i]
+							if not is_location_in_range(performing_player, attack_card, check_position):
+								valid_new_positions.remove_at(i)
+					"anywhere":
+						pass
+					"adjacent_self":
+						for i in range(valid_new_positions.size() - 1, -1, -1):
+							var check_position = valid_new_positions[i]
+							if not (check_position == performing_player.arena_location - 1 or check_position == performing_player.arena_location + 1):
```
Comment (Mach131) [inline scenes/game/local_game.gd:4931]: even if there's no wide card that would use this i'd be more comfortable minimizing where we use the arena_location variable like this; I prefer what you did earlier with checking is_in_location for the spaces next to the potential spike location

File: scenes/game/local_game.gd [hunk 7]
```diff
@@ -4708,6 +4906,163 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
...
+								"destination": destination,
+								"valid_new_positions": valid_new_positions,
+								"and": and_effect
+							}
+						})
+				events += [create_event(Enums.EventType.EventType_ChooseArenaLocationForEffect, performing_player.my_id, 0)]
+			else:
+				# No valid positions to put the buddy, so skip this.
+				# The and effect will occur normally if one exists.
+				pass
```
Comment (Mach131) [inline scenes/game/local_game.gd:5011]: maybe a log message?

File: scenes/game/local_game.gd [hunk 8]
```diff
@@ -4708,6 +4906,163 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
...
+						continue
+					if move_to_opponent and opposing_player.is_in_location(i):
+						valid_new_positions.append(i)
+					else:
+						# Add this space if it is within the amount from the starting buddy location.
+						if abs(location - i) >= move_min and abs(location - i) <= move_max:
+							valid_new_positions.append(i)
+
+				decision_info.limitation.append(location)
+				decision_info.choice.append({
```
Comment (Mach131) [inline scenes/game/local_game.gd:5052]: I assume anything using this effect already checks that you have at least one buddy in play?

File: scenes/game/local_game.gd [hunk 9]
```diff
@@ -5438,7 +5798,10 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
 				}
 				events += [create_event(Enums.EventType.EventType_CardFromHandToGauge_Choice, performing_player.my_id, min_amount, "", max_amount)]
 		"when_hit_force_for_armor":
-			performing_player.strike_stat_boosts.when_hit_force_for_armor = true
+			if 'use_gauge_instead' in effect and effect['use_gauge_instead']:
+				performing_player.strike_stat_boosts.when_hit_force_for_armor = "gauge"
```
Comment (Mach131) [inline scenes/game/local_game.gd:5802]: as mentioned earlier, maybe should have force take priority (don't set "gauge" here if it's already set to "force")

File: scenes/game/local_game.gd [hunk 10]
```diff
@@ -5462,7 +5825,7 @@ func handle_strike_effect(card_id :int, effect, performing_player : Player):
 			assert(false, "ERROR: Unhandled effect type: %s" % effect['effect_type'])
 
 	if not ignore_extra_effects:
-		if "and" in effect:
+		if "and" in effect and effect['and']:
```
Comment (Mach131) [inline scenes/game/local_game.gd:5828]: avoiding setting "and" to null would probably be safer, I think there's at least one other place where I did weird things with the and effect (character actions?)

File: scenes/game/local_game.gd [hunk 11]
```diff
@@ -6159,8 +6526,13 @@ func get_total_power(performing_player : Player, card : GameCard = null):
 	# For now, just assert we're not doing that.
 	assert(performing_player.strike_stat_boosts.power_bonus_multiplier == 1 or performing_player.strike_stat_boosts.power_bonus_multiplier_positive_only == 1)
 
+	var boosted_power = performing_player.strike_stat_boosts.power
+	if performing_player.strike_stat_boosts.power_modify_per_buddy_between:
+		var buddy_count = performing_player.count_buddies_between_opponent()
+		boosted_power += buddy_count * performing_player.strike_stat_boosts.power_modify_per_buddy_between
```
Comment (Mach131) [inline scenes/game/local_game.gd:6532]: hopefully this won't ever have to interact with the positive power multiplier

**Review by Mach131 — CHANGES_REQUESTED** [review]
Buddy logic seems good, from what I can tell! A few missing effects and other comments

File: data/card_definitions.json [hunk 12]
```diff
@@ -15025,6 +15025,392 @@
...
+			"boost_type": "continuous",
+			"force_cost": 0,
+			"cancel_cost": -1,
+			"display_name": "Block And...",
+			"effects": [
+				{
+					"timing": "during_strike",
+					"effect_type": "when_hit_force_for_armor",
+					"use_gauge_instead": true,
+					"amount": 2
```
[the remark this answers is not in the document]
Comment (daniel-k-taylor) [reply data/card_definitions.json:15113]: hmm, right. I think if block is out this should probably just be ignored

File: scenes/game/local_game.gd (hunk 8 above)
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:5052]: yes, that is the assumption

File: scenes/game/local_game.gd (hunk 10 above)
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:5828]: I think it's easier to let us set it to null and just check it here

File: scenes/game/local_game.gd (hunk 11 above)
Comment (daniel-k-taylor) [reply scenes/game/local_game.gd:6532]: please no.
I'll add a comment that it doesn't work.

**Review by Mach131 — APPROVED** [review]

**Pull request merged**